presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -160,96 +163,77 @@ protected CachingHiveMetastore(HiveMetastore delegate, Executor executor, Option
+        partitionStatisticsCache = cacheFactory.createCache(new CacheLoader<>()
+        {
+            @Override
+            public PartitionStatistics load(WithIdentity<HivePartitionName> key)
+            {
+                return loadPartitionColumnStatistics(key);
+            }
 
-        grantedPrincipalsCache = newCacheBuilder(expiresAfterWriteMillis, refreshMills, maximumSize, statsRecording)
-                .build(asyncReloading(CacheLoader.from(this::loadPrincipals), executor));
+            @Override
+            public Map<WithIdentity<HivePartitionName>, PartitionStatistics> loadAll(Iterable<? extends WithIdentity<HivePartitionName>> keys)
+            {
+                return loadPartitionColumnStatistics(keys);
+            }
+        });
+        tableCache = cacheFactory.createCache(this::loadTable);
+        viewNamesCache = cacheFactory.createCache(this::loadAllViews);
+        partitionNamesCache = cacheFactory.createCache(this::loadPartitionNames);
+        partitionFilterCache = cacheFactory.createCache(this::loadPartitionNamesByParts);
+        partitionCache = cacheFactory.createCache(new CacheLoader<>()
+        {
+            @Override
+            public Optional<Partition> load(WithIdentity<HivePartitionName> partitionName)
+            {
+                return loadPartitionByName(partitionName);
+            }
 
-        configValuesCache = newCacheBuilder(expiresAfterWriteMillis, refreshMills, maximumSize, statsRecording)
-                .build(asyncReloading(CacheLoader.from(this::loadConfigValue), executor));
+            @Override
+            public Map<WithIdentity<HivePartitionName>, Optional<Partition>> loadAll(Iterable<? extends WithIdentity<HivePartitionName>> partitionNames)
+            {
+                return loadPartitionsByNames(partitionNames);
+            }
+        });
+        tablePrivilegesCache = cacheFactory.createCache(CacheLoader.from(key -> loadTablePrivileges(key.getDatabase(), key.getTable(), key.getOwner(), key.getPrincipal())));
+        rolesCache = cacheFactory.createCache(this::loadRoles);
+        roleGrantsCache = cacheFactory.createCache(this::loadRoleGrants);
+        grantedPrincipalsCache = cacheFactory.createCache(this::loadPrincipals);
+        configValuesCache = cacheFactory.createCache(this::loadConfigValue);
     }
 
     @Managed
     public void flushCache()
     {
-        databaseNamesCache.invalidateAll();
-        tableNamesCache.invalidateAll();
-        viewNamesCache.invalidateAll();
-        partitionNamesCache.invalidateAll();
-        databaseCache.invalidateAll();
-        tableCache.invalidateAll();
-        partitionCache.invalidateAll();
-        partitionFilterCache.invalidateAll();
-        tablePrivilegesCache.invalidateAll();
-        tableStatisticsCache.invalidateAll();
-        partitionStatisticsCache.invalidateAll();
-        rolesCache.invalidateAll();
+        caches.forEach(Cache::invalidateAll);
```

Comment:
it's nothing new, it's very similar to `Closer` already used in many places, the difference is that here objects can be 'closed' many times

Reply:
Nit: We prefer normal for-each loops. Use `foreach` if there's a specific benefit to doing so, such as the terminal operation of a stream. Note that `Map.forEach` is far superior to for-each as it avoids unpacking the `Entry`.